File: apps/desktop/src/session/types/normalized.ts
/**
 * NormalizedEntry types - must match crates/log-types/src/normalized.rs
 */

export type CommandExitStatus =
  | { type: "exit_code"; code: number }
  | { type: "success"; success: boolean };

export type CommandRunResult = {
  exit_status?: CommandExitStatus | null;
  output?: string | null;
};

export type FileChange =
  | { action: "write"; content: string }
  | { action: "delete" }
  | { action: "rename"; new_path: string }
  | {
      action: "edit";
      unified_diff: string;
      has_line_numbers: boolean;
    };

export type TodoItem = {
  content: string;
  status: string;
  priority?: string | null;
};

/**
 * AskUserQuestion tool input types
 */
export type AskUserQuestionOption = {
  label: string;
  description: string;
};

export type AskUserQuestion = {
  question: string;
  header: string;
  options: AskUserQuestionOption[];
  multiSelect: boolean;
};

export type AskUserQuestionInput = {
  questions: AskUserQuestion[];
};

export type ToolResultValueType = "markdown" | "json";

export type ToolResult = {
  type: ToolResultValueType;
  value: unknown;
};

export type ActionType =
  | { action: "file_read"; path: string }
  | { action: "file_edit"; path: string; changes?: FileChange[] }
  | { action: "command_run"; command: string; result?: CommandRunResult | null }
  | { action: "search"; query: string }
  | { action: "web_fetch"; url: string }
  | {
      action: "tool";
      tool_name: string;
      arguments?: unknown;
      result?: ToolResult | null;
    }
  | { action: "task_create"; description: string }
  | { action: "plan_presentation"; plan: string }
  | { action: "todo_management"; todos: TodoItem[]; operation: string }
  | { action: "other"; description: string };

export type ToolStatus =
  | { status: "created" }
  | { status: "success" }
  | { status: "failed" }
  | { status: "denied"; reason?: string | null }
  | {
      status: "pending_approval";
      approval_id: string;
      requested_at: string;
      timeout_at: string;
    }
  | { status: "timed_out" };

/**
 * Error type classification - matches NormalizedEntryError in Rust
 */
export type NormalizedEntryError =
  | { type: "setup_required" }
  | { type: "other" };

/**
 * Entry type discriminated union - matches NormalizedEntryType in Rust
 */
export type NormalizedEntryType =
  | { type: "user_message" }
  | { type: "user_feedback"; denied_tool: string }
  | { type: "assistant_message" }
  | {
      type: "tool_use";
      tool_name: string;
      action_type: ActionType;
      status: ToolStatus;
    }
  | { type: "system_message" }
  | { type: "error_message"; error_type?: NormalizedEntryError }
  | { type: "thinking" }
  | { type: "loading" }
  | {
      type: "next_action";
      failed: boolean;
      execution_processes: number;
      needs_setup: boolean;
    };

/**
 * A normalized entry representing a single piece of agent conversation.
 * The `id` field is added client-side for React keys.
 */
export type NormalizedEntry = {
  id: string;
  timestamp?: string | null;
  entry_type: NormalizedEntryType;
  content: string;
  metadata?: Record<string, unknown>;
};

/**
 * Display entry type with key.
 * Keeps NORMALIZED_ENTRY, STDOUT, STDERR as separate types for proper rendering.
 */
export type DisplayEntry =
  | { type: "NORMALIZED_ENTRY"; content: NormalizedEntry; key: string }
  | { type: "STDOUT"; content: string; key: string }
  | { type: "STDERR"; content: string; key: string };

File: crates/log-types/src/normalized.rs
//! Normalized log entry types for structured agent output display.
//!
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Result type for tool execution values.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ToolResultValueType {
    Markdown,
    Json,
}

/// Tool result with typed value.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ToolResult {
    #[serde(rename = "type")]
    pub result_type: ToolResultValueType,
    /// For Markdown, this will be a JSON string; for JSON, a structured value
    pub value: Value,
}

impl ToolResult {
    pub fn markdown<S: Into<String>>(markdown: S) -> Self {
        Self {
            result_type: ToolResultValueType::Markdown,
            value: Value::String(markdown.into()),
        }
    }

    pub fn json(value: Value) -> Self {
        Self {
            result_type: ToolResultValueType::Json,
            value,
        }
    }
}

/// Exit status of a command.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum CommandExitStatus {
    ExitCode { code: i32 },
    Success { success: bool },
}

/// Result of a command execution.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CommandRunResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub exit_status: Option<CommandExitStatus>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<String>,
}

/// Error type classification.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NormalizedEntryError {
    SetupRequired,
    Other,
}

/// The type of normalized entry.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum NormalizedEntryType {
    /// User input message
    UserMessage,
    /// User feedback (denial reason, etc.)
    UserFeedback { denied_tool: String },
    /// Assistant response text
    AssistantMessage,
    /// Tool use action
    ToolUse {
        tool_name: String,
        action_type: ActionType,
        status: ToolStatus,
    },
    /// System message (initialization, model info, etc.)
    SystemMessage,
    /// Error message
    ErrorMessage { error_type: NormalizedEntryError },
    /// Thinking/reasoning block
    Thinking,
    /// Loading state
    Loading,
    /// Next action summary
    NextAction {
        failed: bool,
        execution_processes: usize,
        needs_setup: bool,
    },
}

/// A normalized entry representing a single piece of agent conversation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct NormalizedEntry {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub timestamp: Option<String>,
    #[serde(rename = "type")]
    pub entry_type: NormalizedEntryType,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<Value>,
}

/// Content patterns whose `SystemMessage` entries must never reach the UI.
///
/// These are agent-CLI protocol events (e.g. Claude Code's
/// `{"type":"system","subtype":"status"}`) that normalize into
/// `SystemMessage` but carry no user-facing value. Suppression happens at
/// the normalizer boundary so the entries never enter the patch stream,
/// the conversation store, or the render path.
pub const SUPPRESSED_SYSTEM_CONTENT_PATTERNS: &[&str] = &["System: status"];

/// Returns true when a would-be `SystemMessage` content matches a known
/// suppression pattern and the entry must not be emitted.
///
/// Matching is whitespace-trimmed and requires either an exact match or a
/// word-boundary prefix (pattern followed by whitespace). This avoids
/// accidentally swallowing content like "System: statusbar".
pub fn should_suppress_system_message(content: &str) -> bool {
    let trimmed = content.trim();
    SUPPRESSED_SYSTEM_CONTENT_PATTERNS.iter().any(|pattern| {
        if trimmed == *pattern {
            return true;
        }
        trimmed
            .strip_prefix(*pattern)
            .is_some_and(|rest| rest.starts_with(|c: char| c.is_whitespace()))
    })
}

/// Status of a tool execution.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum ToolStatus {
    /// Tool call created, awaiting result
    #[default]
    Created,
    /// Tool executed successfully
    Success,
    /// Tool execution failed
    Failed,
    /// Tool was denied/skipped
    Denied {
        #[serde(skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
    },
    /// Waiting for approval
    PendingApproval {
        approval_id: String,
        requested_at: String,
        timeout_at: String,
    },
    /// Approval timed out
    TimedOut,
}

/// A TODO item from TodoWrite.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TodoItem {
    pub content: String,
    pub status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub priority: Option<String>,
}

/// Describes what action a tool is performing.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum ActionType {
    /// Reading a file
    FileRead { path: String },
    /// Editing/writing a file
    FileEdit {
        path: String,
        #[serde(default, skip_serializing_if = "Vec::is_empty")]
        changes: Vec<FileChange>,
    },
    /// Running a command
    CommandRun {
        command: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        result: Option<CommandRunResult>,
    },
    /// Searching (grep, glob, etc.)
    Search { query: String },
    /// Fetching a web page
    WebFetch { url: String },
    /// Generic tool with optional arguments and result for rich rendering
    Tool {
        tool_name: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        arguments: Option<Value>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        result: Option<ToolResult>,
    },
    /// Creating a subtask
    TaskCreate { description: String },
    /// Presenting a plan (ExitPlanMode)
    PlanPresentation { plan: String },
    /// TODO list management
    TodoManagement {
        todos: Vec<TodoItem>,
        operation: String,
    },
    /// Other/unknown action
    Other { description: String },
}

/// Represents a file change (edit or write).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum FileChange {
    /// Create a file if it doesn't exist, and overwrite its content.
    Write { content: String },
    /// Delete a file.
    Delete,
    /// Rename a file.
    Rename { new_path: String },
    /// Edit a file with a unified diff.
    Edit {
        /// Unified diff containing file header and hunks.
        unified_diff: String,
        /// Whether line numbers in the hunks are reliable.
        has_line_numbers: bool,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn suppresses_exact_system_status() {
        assert!(should_suppress_system_message("System: status"));
    }

    #[test]
    fn suppresses_trimmed_system_status() {
        assert!(should_suppress_system_message("  System: status\n"));
    }

    #[test]
    fn suppresses_system_status_with_trailing_detail() {
        assert!(should_suppress_system_message("System: status running"));
    }

    #[test]
    fn does_not_suppress_unrelated_system_messages() {
        assert!(!should_suppress_system_message("Run with Opus 4.7"));
        assert!(!should_suppress_system_message("System: init"));
        assert!(!should_suppress_system_message("System message"));
        assert!(!should_suppress_system_message(""));
    }

    #[test]
    fn does_not_suppress_word_boundary_false_positive() {
        assert!(!should_suppress_system_message("System: statusbar"));
    }
}
